src/trust/chain.ts

```diff
@@ -0,0 +1,109 @@
+import {
+  decode as decodeJwt,
+  verify as verifyJwt,
+} from "@pagopa/io-react-native-jwt";
+import {
+  EntityConfiguration,
+  EntityStatement,
+  TrustAnchorEntityConfiguration,
+} from "./types";
+import { JWK } from "../utils/jwk";
+import { IoWalletError } from "../utils/errors";
+import * as z from "zod";
+import type { JWTDecodeResult } from "@pagopa/io-react-native-jwt/lib/typescript/types";
+
+type ParsedToken = {
+  header: JWTDecodeResult["protectedHeader"];
+  payload: JWTDecodeResult["payload"];
+};
+
+// Verify a token signature
+// The kid is extracted from the token header
+const verify = async (
+  token: string,
+  kid: string,
+  jwks: JWK[]
+): Promise<ParsedToken> => {
+  const jwk = jwks.find((k) => k.kid === kid);
+  if (!jwk) {
+    throw new Error(`Invalid kid: ${kid}, token: ${token}`);
+  }
+  const { protectedHeader: header, payload } = await verifyJwt(token, jwk);
+  return { header, payload };
+};
+
+const decode = (token: string) => {
+  const { protectedHeader: header, payload } = decodeJwt(token);
+  return { header, payload };
+};
+
+// The first element of the chain is supposed to be the Entity Configuration for the document issuer
+const FirstElementShape = EntityConfiguration;
+// Each element but the first is supposed to be an Entity Statement
+const MiddleElementShape = EntityStatement;
+// The last element of the chain can either be an Entity Statement
+//  or the Entity Configuration for the known Trust Anchor
+const LastElementShape = z.union([
+  EntityStatement,
+  TrustAnchorEntityConfiguration,
+]);
```

Comment:
What do you think about making the control programmatic? A priori we do not know how many entities there are in the chain. It could be 2,3 or 4.

Reply:
I don't understand the question. Those are utility types meant for a better undestand of chain structure. It allows for chains of any shape like

```ts
[FirstElementShape, LastElementShape]
[FirstElementShape, MiddleElementShape, LastElementShape]
[FirstElementShape, MiddleElementShape, MiddleElementShape, LastElementShape]
[FirstElementShape, MiddleElementShape, MiddleElementShape, MiddleElementShape, LastElementShape]
[FirstElementShape, MiddleElementShape, ... , MiddleElementShape, LastElementShape]
... and so on
```

The control is then programmatic, as it iterates through the actual chain to be verified, expecting each element to be in shape according to its position in the chain. (see `selectTokenShape`)